content/2019/a-test-to-attest-to/index.md

```diff
@@ -0,0 +1,1183 @@
+add reports to the special `v:errors` list.
+
+That is:
+
+> A test is considered to have failed if `v:errors` is a non-empty list.
+
+Let's explore the other sources of "failure" too.
+
+### Exceptions and tracebacks
+
+Errors and exceptions do occur, and they can be trapped with `:try` etc. For
+reasons that will become clearer when we wrap this all up into a neat little
+"framework", it is not considered good practice to "fail" a test by throwing or
+triggering an uncaught error/exception. It's best to use the assert functions,
+or add things to `v:errors` with `call add( v:errors, ... )`.
+
+However, an exception might be thrown due to a bug in our code! We should catch
+these uncaught exceptions and mark our test as failed if there are any:
+
+> A test is considered to have failed if an exception is thrown and not caught
+> within the test.
+
+```vim
+try
+  " do the test
+catch
+  call assert_report( "Uncaught exception in test: "
+                    \ . v:exception 
+                    \ . " at " 
+                    \ . v:throwpoint )
+endtry
+```
+
+### Early exit
+
+While it probably goes without saying, it's not a good idea for the test to
+cause Vim to exit. So, we want to trap that and report it as a
+failure. For that we can use [`VimLeavePre`][h-vimleavepre] autocommand:
+
+```vim
+au VimLeavePre * call s:EarlyExit()
+try
+  " do the test
+catch
+  " Handle uncaught
+finally
+  au! VimLeavePre
+endtry
+```
+
+The `s:EarlyExit()` function will be defined later, but you can guess that it
+adds something to `v:errors` and then quits with an error code.
+
+## Reporting results
+
+After running your test, you need to know if it was successful, and if not, why
+not. The former is fairly easy: we can make Vim exit with a nonzero exit code on
+failure. For this, there is [`:quit!`][h-quit] for success and
+[`:cquit!`][h-cquit] for failure:
+
+```vim
+if len( v:errors ) > 0
+  cquit!
+else
+  quit!
+endif
+```
+
+Unfortunately, getting the _reason_ for failure is not so straightforward. Due
+to Vim being a terminal-mode application itself, the obvious choice 
+(print something to stdout/stderr) isn't really available. `:echom` and suchlike
+will all be lost when Vim exits. So what we normally do is to write all the
+messages reported to a file (called `<file>.failed.log`, why not) and have a
+wrapper script detect the failure and print them out.
+
+```vim
+" Append errors to test failure log
+let logfile = expand( "<sfile>:p:t" ) . ".failed.log"
+call writefile( v:errors, logfile, 'as' )
+```
+
+## Putting it all together
+
+So, recalling that the test we want to run is:
+
+```vim
+call feedkeys( "i\<C-x>\<C-u>", 'xt' )
+call assert_equal( 'Jan', getline( 1 ) )
+```
+
+We can create [a script to automate this for CompleteSimple][script-simple],
+encompassing all of the above, as follows:
+
+```vim
+let init_script = expand( '<sfile>:p:h' ) . '/../support/test_simple.vim'
+execute 'source ' . init_script
```

Comment:
Could be just `execute 'source' init_script`

Reply:
Also true. I probably won't change this as it's not really material to the article.